ipalib/install/certmonger.py

```diff
@@ -317,10 +317,62 @@ def request_and_wait_for_cert(
     state = wait_for_request(reqId, api.env.startup_timeout)
     ca_error = get_request_value(reqId, 'ca-error')
     if state != 'MONITORING' or ca_error:
-        raise RuntimeError("Certificate issuance failed ({})".format(state))
+        raise RuntimeError(
+            "Certificate issuance failed ({}: {})".format(state, ca_error)
+        )
+
     return reqId
 
 
+def request_and_retry_cert(
+        certpath, subject, principal, nickname=None, passwd_fname=None,
+        dns=None, ca='IPA', profile=None,
+        pre_command=None, post_command=None, storage='NSSDB', perms=None,
+        timeout=120):
+    """Request certificate, wait and resubmit failing requests
+
+    The function is similar to request_and_wait_for_cert, but also resubmits
+    failing cert requests. During parallel replica installation, a request
+    sometimes fails with CA_REJECTED or CA_UNREACHABLE. The error occur when
+    the master is either busy or some information haven't been replicated
+    yet. Even a stuck request can be recovered, e.g. when permission and
+    group information have been replicated.
+    """
+    req_id = request_cert(
+        certpath, subject, principal, nickname, passwd_fname, dns, ca,
+        profile, pre_command, post_command, storage, perms
+    )
+
+    deadline = time.time() + timeout
+    while True:
```

Comment:
My first implementation looked like that. But I didn't want to define ``state`` and ``ca_error``.

Reply:
Two more reasons why I prefer ``while True`` loop here:

* The current implementation is basically a ``do/while`` loop. It will run at least one time with ``timeout=0``.
* The loop has three exit conditions: success, timeout, and error. I find it confusing to have only one exit check in the loop header.

I can certainly add a comment and make the code flow more obvious.